File: Back-End/backend/src/services/tagRiskService.ts
import { Tag } from '../models/Tag';
import { AssetTag } from '../models/AssetTag';
import { Holding } from '../models/Holding';
import { Stock } from '../models/Stock';
import { MarketStateName } from '../models/MarketState';

export interface RecalculateResult {
  updatedCount: number;
  skippedCount: number;
}

function clamp(v: number): number {
  return Math.max(0, Math.min(3, v));
}

function popStd(values: number[]): number {
  if (values.length < 2) return 0;
  const mean = values.reduce((a, b) => a + b, 0) / values.length;
  return Math.sqrt(values.reduce((s, v) => s + (v - mean) ** 2, 0) / values.length);
}

function dailyReturns(closes: number[]): number[] {
  const out: number[] = [];
  for (let i = 1; i < closes.length; i++) {
    if (closes[i - 1] > 0) out.push((closes[i] - closes[i - 1]) / closes[i - 1]);
  }
  return out;
}

export async function recalculateDynamicRisk(
  marketState: MarketStateName
): Promise<RecalculateResult> {
  const [tags, allAssetTags, allHoldings] = await Promise.all([
    Tag.findAll(),
    AssetTag.findAll(),
    Holding.findAll(),
  ]);

  const activeSet = new Set(allHoldings.filter(h => h.sharesHeld > 0).map(h => h.stockId));

  // tagName → 有效持股清單（只含 sharesHeld > 0 的股票）
  const tagHoldingsMap = new Map<string, { stockCode: string; weightRatio: number }[]>();
  for (const at of allAssetTags) {
    if (!activeSet.has(at.stockCode)) continue;
    const list = tagHoldingsMap.get(at.tagName) ?? [];
    list.push({ stockCode: at.stockCode, weightRatio: at.weightRatio });
    tagHoldingsMap.set(at.tagName, list);
  }

  // 收集需要歷史資料的唯一股票代號
  const neededStocks = new Set<string>();
  for (const holdings of tagHoldingsMap.values()) {
    for (const { stockCode } of holdings) neededStocks.add(stockCode);
  }

  // 並行取得 90 日收盤價，單筆失敗不中斷
  const historyResults = await Promise.allSettled(
    [...neededStocks].map(async (stockCode) => ({
      stockCode,
      closes: (await Stock.getHistory(stockCode, 90)).map(p => p.close),
    }))
  );

  const closesMap = new Map<string, number[]>();
  for (const r of historyResults) {
    if (r.status === 'fulfilled') closesMap.set(r.value.stockCode, r.value.closes);
  }

  // 計算各 Tag 的 vol_ratio → presets → dynamicRisk
  const updates: {
    id: string;
    dynamicRisk: number;
    marketStatePresets: { riskOn: number; riskOff: number; liquidityDry: number };
  }[] = [];
  let skippedCount = 0;

  for (const tag of tags) {
    const holdings = tagHoldingsMap.get(tag.name);
    if (!holdings || holdings.length === 0) {
      skippedCount++;
      continue;
    }

    // 建立各股日報酬序列（排除取不到歷史資料的股票）
    const stockSeries: { returns: number[]; weight: number }[] = [];
    for (const { stockCode, weightRatio } of holdings) {
      const closes = closesMap.get(stockCode);
      if (!closes || closes.length < 2) continue;
      stockSeries.push({ returns: dailyReturns(closes), weight: weightRatio / 100 });
    }

    if (stockSeries.length === 0) {
      skippedCount++;
      continue;
    }

    // 對齊最短序列，計算加權 Tag 日報酬
    const minLen = Math.min(...stockSeries.map(s => s.returns.length));
    const tagReturns: number[] = [];
    for (let i = 0; i < minLen; i++) {
      let r = 0;
      for (const { returns, weight } of stockSeries) r += weight * returns[i];
      tagReturns.push(r);
    }

    // vol_ratio：資料不足 20 筆或 baseVol = 0 時預設 1.0
    let volRatio = 1.0;
    if (tagReturns.length >= 20) {
      const recentVol = popStd(tagReturns.slice(-20));
      const baseVol   = popStd(tagReturns);
      if (baseVol > 0) volRatio = recentVol / baseVol;
    }

    const { baseRisk } = tag;
    const riskOn       = clamp(baseRisk * 1.3 * volRatio);
    const riskOff      = clamp(baseRisk * 1.8 * volRatio);
    const liquidityDry = clamp(baseRisk * 2.5 * volRatio);

    const dynamicRisk =
      marketState === 'risk-on'       ? riskOn       :
      marketState === 'risk-off'      ? riskOff      :
      marketState === 'liquidity-dry' ? liquidityDry :
      clamp(baseRisk * volRatio); // neutral

    updates.push({ id: tag.id, dynamicRisk, marketStatePresets: { riskOn, riskOff, liquidityDry } });
  }

  if (updates.length > 0) await Tag.batchUpdateRisk(updates);

  return { updatedCount: updates.length, skippedCount };
}

File: Back-End/python-backend/services/tag_risk_service.py
"""動態風險重算服務（忠實對應 Node.js tagRiskService.ts）"""
from __future__ import annotations
from core.executors import get_executor
from services.firestore import get_db
from services.yahoo_finance import get_history_closes


# ─── 數學工具 ──────────────────────────────────────────────────────────────────

def _pop_std(values: list[float]) -> float:
    if len(values) < 2:
        return 0.0
    mean = sum(values) / len(values)
    return (sum((v - mean) ** 2 for v in values) / len(values)) ** 0.5


def _daily_returns(closes: list[float]) -> list[float]:
    out = []
    for i in range(1, len(closes)):
        if closes[i - 1] > 0:
            out.append((closes[i] - closes[i - 1]) / closes[i - 1])
    return out


def _clamp(v: float) -> float:
    return max(0.0, min(3.0, v))


def _r2(v: float) -> float:
    return round(v, 2)


# ─── 核心計算 ──────────────────────────────────────────────────────────────────

def recalculate_dynamic_risk(market_state: str) -> dict:
    db = get_db()

    # 讀取 tags / asset_tags / holdings
    tags_snap      = db.collection("tags").order_by("name").get()
    asset_tags_snap = db.collection("asset_tags").get()
    holdings_snap  = db.collection("holdings").get()

    tags = []
    for doc in tags_snap:
        d = doc.to_dict()
        msp_raw = d.get("market_state_presets")
        msp = None
        if msp_raw:
            msp = {
                "riskOn":       msp_raw.get("risk_on"),
                "riskOff":      msp_raw.get("risk_off"),
                "liquidityDry": msp_raw.get("liquidity_dry"),
            }
        tags.append({
            "id":       doc.id,
            "name":     d.get("name"),
            "baseRisk": d.get("base_risk", 0),
            "marketStatePresets": msp,
        })

    # 活躍持股 set（sharesHeld > 0）
    active_set = {
        doc.to_dict().get("stock_id", doc.id)
        for doc in holdings_snap
        if (doc.to_dict().get("shares_held") or 0) > 0
    }

    # tagName → [{stockCode, weightRatio}]
    tag_holdings_map: dict[str, list[dict]] = {}
    for doc in asset_tags_snap:
        d = doc.to_dict()
        stock_code = d.get("stock_code", "")
        if stock_code not in active_set:
            continue
        tag_name = d.get("tag_name", "")
        lst = tag_holdings_map.setdefault(tag_name, [])
        lst.append({"stockCode": stock_code, "weightRatio": d.get("weight_ratio", 0)})

    # 收集需要歷史資料的股票
    needed = set()
    for items in tag_holdings_map.values():
        for item in items:
            needed.add(item["stockCode"])

    # 並行取得 90 日收盤價（使用共用 executor）
    closes_map: dict[str, list[float]] = {}
    if needed:
        executor = get_executor()
        futures = {executor.submit(get_history_closes, sc): sc for sc in needed}
        for fut, sc in futures.items():
            try:
                closes = fut.result()
                if closes:
                    closes_map[sc] = closes
            except Exception:
                pass

    # 計算各 Tag 的 vol_ratio → presets → dynamicRisk
    updates = []
    skipped = 0

    for tag in tags:
        holdings = tag_holdings_map.get(tag["name"], [])
        if not holdings:
            skipped += 1
            continue

        stock_series = []
        for item in holdings:
            closes = closes_map.get(item["stockCode"])
            if not closes or len(closes) < 2:
                continue
            stock_series.append({
                "returns": _daily_returns(closes),
                "weight":  item["weightRatio"] / 100,
            })

        if not stock_series:
            skipped += 1
            continue

        min_len = min(len(s["returns"]) for s in stock_series)
        tag_returns = []
        for i in range(min_len):
            r = sum(s["weight"] * s["returns"][i] for s in stock_series)
            tag_returns.append(r)

        vol_ratio = 1.0
        if len(tag_returns) >= 20:
            recent_vol = _pop_std(tag_returns[-20:])
            base_vol   = _pop_std(tag_returns)
            if base_vol > 0:
                vol_ratio = recent_vol / base_vol

        base_risk    = tag["baseRisk"]
        risk_on      = _r2(_clamp(base_risk * 1.3 * vol_ratio))
        risk_off     = _r2(_clamp(base_risk * 1.8 * vol_ratio))
        liquidity_dry = _r2(_clamp(base_risk * 2.5 * vol_ratio))

        if market_state == "risk-on":
            dynamic_risk = risk_on
        elif market_state == "risk-off":
            dynamic_risk = risk_off
        elif market_state == "liquidity-dry":
            dynamic_risk = liquidity_dry
        else:
            dynamic_risk = _r2(_clamp(base_risk * vol_ratio))

        updates.append({
            "id":          tag["id"],
            "dynamicRisk": dynamic_risk,
            "marketStatePresets": {
                "riskOn":       risk_on,
                "riskOff":      risk_off,
                "liquidityDry": liquidity_dry,
            },
        })

    if updates:
        batch = db.batch()
        for u in updates:
            ref = db.collection("tags").document(u["id"])
            batch.update(ref, {
                "dynamic_risk": u["dynamicRisk"],
                "market_state_presets": {
                    "risk_on":       u["marketStatePresets"]["riskOn"],
                    "risk_off":      u["marketStatePresets"]["riskOff"],
                    "liquidity_dry": u["marketStatePresets"]["liquidityDry"],
                },
            })
        batch.commit()

    return {"updatedCount": len(updates), "skippedCount": skipped}
